File: 2015/dia03/main.py
input = open("input.txt", "r").read().strip()
# print(input)

instruction_map = {
    ">": (1, 0),
    "<": (-1, 0),
    "^": (0, 1),
    "v": (0, -1),
}


def parte_1():
    position = (0, 0)
    positions = set([position])
    for instruction in input:
        dx, dy = instruction_map[instruction]
        x, y = position
        position = (x + dx, y + dy)
        positions.add(position)
    print(len(positions))


def parte_2():
    santa_position = (0, 0)
    robo_santa_position = (0, 0)
    positions = set([santa_position, robo_santa_position])
    flag = True
    for instruction in input:
        dx, dy = instruction_map[instruction]
        if flag:
            x, y = santa_position
            santa_position = (x + dx, y + dy)
            positions.add(santa_position)
        else:
            x, y = robo_santa_position
            robo_santa_position = (x + dx, y + dy)
            positions.add(robo_santa_position)
        flag = not flag
    print(len(positions))


if __name__ == "__main__":
    parte_1()
    parte_2()

File: 2015/dia03/cpp/main.cpp
#include <algorithm>
#include <boost/functional/hash.hpp>
#include <fstream>
#include <functional>
#include <numeric>
#include <print>
#include <ranges>
#include <unordered_set>
#include <vector>

namespace rg = std::ranges;
namespace vw = std::views;

void process_walk(char c, std::tuple<int, int>& position) {
    if (c == '^') std::get<1>(position)++;
    if (c == 'v') std::get<1>(position)--;
    if (c == '>') std::get<0>(position)++;
    if (c == '<') std::get<0>(position)--;
}

void parte_1(std::vector<std::string>& lines) {
    std::string instructions = lines[0];
    std::tuple<int, int> position = {0, 0};
    std::unordered_set<std::tuple<int, int>, boost::hash<std::tuple<int, int>>>
        visited_houses;
    visited_houses.insert(position);
    for (char c : instructions) {
        process_walk(c, position);
        visited_houses.insert(position);
    }
    std::print("Total parte 1: {}\n", visited_houses.size());
};

struct boost_hash_tuple {
    template <typename T>
    size_t operator()(const std::tuple<T, T>& x) const {
        return boost::hash_value(x);
    }
};

void parte_2(std::vector<std::string>& lines) {
    std::string instructions = lines[0];
    std::tuple<int, int> santa_position = {0, 0};
    std::tuple<int, int> robo_santa_position = {0, 0};
    std::unordered_set<std::tuple<int, int>, boost_hash_tuple> visited_houses;
    visited_houses.insert(santa_position);
    int flag{0};
    for (char c : instructions) {
        if (flag) {
            flag = 0;
            process_walk(c, santa_position);
            visited_houses.insert(santa_position);
        } else {
            flag = 1;
            process_walk(c, robo_santa_position);
            visited_houses.insert(robo_santa_position);
        }
    }
    std::print("Total parte 2: {}\n", visited_houses.size());
};

int main(int, char** argv) {
    std::print("\n");

    std::ifstream input;
    input.open(argv[1]);

    std::string line;
    std::vector<std::string> lines;
    while (std::getline(input, line)) {
        lines.push_back(line);
    }
    input.close();

    parte_1(lines);
    parte_2(lines);

    return 0;
}
